controller/telemetry/server.go

```diff
@@ -184,29 +184,52 @@ func NewServer(addr, prometheusUrl string, ignoredNamespaces []string, kubeconfi
 func (s *server) Query(ctx context.Context, req *read.QueryRequest) (*read.QueryResponse, error) {
 	log.Debugf("Query request: %+v", req)
 
-	start := time.Unix(0, req.StartMs*int64(time.Millisecond))
-	end := time.Unix(0, req.EndMs*int64(time.Millisecond))
+	var res model.Value
+	var err error
 
-	step, err := time.ParseDuration(req.Step)
-	if err != nil {
-		log.Errorf("ParseDuration(%+v) failed with: %+v", req.Step, err)
-		return nil, err
-	}
+	if req.StartMs != 0 && req.EndMs != 0 && req.Step != "" {
+		start := time.Unix(0, req.StartMs*int64(time.Millisecond))
+		end := time.Unix(0, req.EndMs*int64(time.Millisecond))
+		step, err := time.ParseDuration(req.Step)
+		if err != nil {
+			log.Errorf("ParseDuration(%+v) failed with: %+v", req.Step, err)
+			return nil, err
+		}
 
-	queryRange := v1.Range{Start: start, End: end, Step: step}
-	res, err := s.prometheusApi.QueryRange(ctx, req.Query, queryRange)
-	if err != nil {
-		log.Errorf("QueryRange(%+v, %+v) failed with: %+v", req.Query, queryRange, err)
-		return nil, err
-	}
+		queryRange := v1.Range{
+			Start: start,
+			End:   end,
+			Step:  step,
+		}
 
-	if res.Type() != model.ValMatrix {
-		return nil, fmt.Errorf("Unexpected query result type: %s", res.Type())
+		res, err = s.prometheusAPI.QueryRange(ctx, req.Query, queryRange)
+		if err != nil {
+			log.Errorf("QueryRange(%+v, %+v) failed with: %+v", req.Query, queryRange, err)
+			return nil, err
+		}
+	} else {
+		res, err = s.prometheusAPI.Query(ctx, req.Query, time.Time{})
+		if err != nil {
+			log.Errorf("Query(%+v, %+v) failed with: %+v", req.Query, time.Time{}, err)
+			return nil, err
+		}
 	}
+	log.Debugf("Query response: %+v", res)
 
 	samples := make([]*read.Sample, 0)
-	for _, s := range res.(model.Matrix) {
-		samples = append(samples, convertSampleStream(s))
```

Comment:
this is not specific to this change, but I'm struggling a bit to understand the API.  Is a "Sample" a timeseries?  and each sample contains the data points for that timeseries?  and in summary mode each Sample will only have one data point?

Reply:
correct, a `read.Sample` represents a single metric over time, and, in summary mode `read.Sample` will contain one data point. this is another artifact of telemetry's `Query()` endpoint supporting both timeseries and non-timeseries data (so it always returns a timeseries-like data structure). this is another area to clean up in #312.